### ISSUE TITLE
Paper unifying the Normalized, Adaptive and Invariant Update Equations ?

### ISSUE THREAD

**Issue by apoorvreddy** [JohnLangford/vowpal_wabbit#1204]
Hi @JohnLangford @arielf 

I am attempting to understand the SGD update rules in VW by going through the source code and tried to draw parallels between the code and the following papers:
1) [Normalized online learning](http://www.cs.cmu.edu/~sross1/publications/uai13_normalized.pdf) - This one captures the normalized and adaptive updates part, and
2) [Online Importance Weight Aware Updates](https://arxiv.org/pdf/1011.1576.pdf) - This one captures the invariant updates part.

### Update Rules in the Paper
1) [Normalized online learning](http://www.cs.cmu.edu/~sross1/publications/uai13_normalized.pdf) - 

<img width="301" alt="screen shot 2017-02-27 at 11 49 15 am" src="https://cloud.githubusercontent.com/assets/9053270/23350762/d4d0f98e-fce2-11e6-803c-8364d7bc3be2.png">

2) [Online Importance Weight Aware Updates](https://arxiv.org/pdf/1011.1576.pdf)
<img width="337" alt="screen shot 2017-02-27 at 11 46 18 am" src="https://cloud.githubusercontent.com/assets/9053270/23350711/81740006-fce2-11e6-8c09-b2cb69f66598.png">

### Actual Update rule in VW:
<img width="422" alt="screen shot 2017-02-27 at 4 46 03 pm" src="https://cloud.githubusercontent.com/assets/9053270/23359471/61a731ec-fd0c-11e6-8cb7-bf40625c6bee.png">

### Differences between the Papers and the VW code:
1) G_i is accumulating the product of the squared gradients and the squared feature values ? Is there any reason for this ?
2) I am interested in the neural network code in VW. I've noticed that N is shared across both the output and hidden layers in the network ? Is there any paper which talks about this ?

**Comment by arielf** [comment]
Sorry, I'm not familiar enough with the update calculation details (in both the code and the papers) to answer at this level of detail.  One thing I could think about is that in real-life code you sometimes have to account for more than the classic/theoretical case because of practical reasons. These include avoiding floating-point exceptions, accounting for various user-supplied options, etc.

As for the `nn` code: it was written by Paul Mineiro, so you may want to ask him directly.

**Comment by apoorvreddy** [comment]
CC @pmineiro :)

**Comment by pmineiro** [comment]
If there are quantities like the "total example weight" or "total sum of example norms" accumulated in gd.cc into (effectively) globals, this might be considered a bug from the reduction framework standpoint, since it means the reduced-to predictors are coupled.

Wrt nn specifically, I was (incorrectly?) thinking of the reduced-to predictors as completely distinct.

**Comment by apoorvreddy** [comment]
The sum of example norms N for one example, is updated here in [Line 460](https://github.com/JohnLangford/vowpal_wabbit/blob/ee2bcac9354ae91b97653b86024a24b41f6c7bb2/vowpalwabbit/gd.cc#L460) in the method [pred_per_update_feature](https://github.com/JohnLangford/vowpal_wabbit/blob/ee2bcac9354ae91b97653b86024a24b41f6c7bb2/vowpalwabbit/gd.cc#L433) which is called per feature [here in get_pred_per_update](https://github.com/JohnLangford/vowpal_wabbit/blob/ee2bcac9354ae91b97653b86024a24b41f6c7bb2/vowpalwabbit/gd.cc#L477)

This N is then added to the global sum of example norms [here in get_pred_per_update](https://github.com/JohnLangford/vowpal_wabbit/blob/ee2bcac9354ae91b97653b86024a24b41f6c7bb2/vowpalwabbit/gd.cc#L481)

The flow of execution for output layer update:
[base.learn](https://github.com/JohnLangford/vowpal_wabbit/blob/ee2bcac9354ae91b97653b86024a24b41f6c7bb2/vowpalwabbit/nn.cc#L269) -> [update](https://github.com/JohnLangford/vowpal_wabbit/blob/ee2bcac9354ae91b97653b86024a24b41f6c7bb2/vowpalwabbit/gd.cc#L549) -> compute_update -> sensitivity -> pred_per_update

The flow of execution for hidden layer update:
[base.update](https://github.com/JohnLangford/vowpal_wabbit/blob/ee2bcac9354ae91b97653b86024a24b41f6c7bb2/vowpalwabbit/nn.cc#L313) -> update -> compute_update -> sensitivity -> pred_per_update

Here base is the base learner which is the gd learner.

Also, I believe, that each of the hidden nodes should have a corresponding N as well, i.e. if there are k hidden nodes in the network, there must be k + 1 Ns (1 for the output node and k for the hidden nodes).

Thoughts ?

**Comment by apoorvreddy** [comment]
Hi @pmineiro and @JohnLangford ! What is a workaround for this ?

**Comment by pmineiro** [comment]
There's no workaround.  Are you experiencing poor performance on some task?

**Comment by JohnLangford** [comment]
W.r.t. your original questions:
1) I think you are misreading the code---there is a part of the gradient due to the loss function and a part due to the feature value.  These are computed separately for computational reasons.   I expect you you are thinking of the first part as the full gradient while it's only handling to loss function related part.
(2) There is coupling.  This could be removed by storing a vector of norms, one per base predictor.  That's not currently done, but it seems like it's plausibly a good idea when the different base predictors are used in very different ways.  Have you noticed an issue with this?

**Comment by apoorvreddy** [comment]
Hi John,

Thank you for replying. I will have to go through the code again regarding comment 1.

Regarding comment 2, I felt that normalized update was incorrectly implemented, and hence I felt that an issue should be raised. Currently, I do not know if I can achieve better performance if every base predictor has a separate norm. Do you have any suggestion as to the best way I can implement this and possibly raise a PR ?

**Comment by JohnLangford** [comment]
Sure.  This is the critical function: https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/gd.cc#L110

You need to vectorize all the variables to depend on ec.ft_offset https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/example.h#L90  I'd also recommend pulling the global variable (in g.all->...) into the local datastructure.
